README.md
[![ContentDB](https://content.minetest.net/packages/amalon/numeracy/shields/title/)](https://content.minetest.net/packages/amalon/numeracy/)
[![ContentDB](https://content.minetest.net/packages/amalon/numeracy/shields/downloads/)](https://content.minetest.net/packages/amalon/numeracy/)

Numeracy Mod for Minetest
=========================

Adds blocks for exploring numeracy with children, which join together and
change appearance depending on the number of connected blocks. Inspired by the
kids TV programme Numberblocks.

![Screenshot](docs/screenshot.jpg)

Version: 0.4

Licence: LGPL 2.1 or later, CC-BY-SA 3.0 UNPORTED (see LICENSE.txt)

Dependencies: none

 - [ContentDB](https://content.minetest.net/packages/amalon/numeracy/)
 - [Forum topic](https://forum.minetest.net/viewtopic.php?t=26516)
 - [Report bugs](https://github.com/amalon/minetest-numeracy/issues)
 - [Matrix room](https://matrix.to/#/#minetest-numeracy:hoganfam.uk?via=hoganfam.uk)

Installation
------------

Unzip the archive and place it in minetest/mods/minetest/

(  Linux: If you have a linux system-wide installation place
	it in ~/.minetest/mods/minetest/.  )

(  If you only want this to be used in a single world, place
	the folder in worldmods/ in your worlddirectory.  )

For further information or help see:
http://wiki.minetest.com/wiki/Installing_Mods

Getting Started
---------------

Get a "Numeracy Block" from the creative inventry and start placing them down
in groups. Watch the blocks change appearance and the number above the blocks
change.

Try measuring the distance between two places by placing a trail of blocks
down.

Try making as many different rectangles or other shapes out of the same number
of blocks.

Try exploring sequences of natural numbers, times tables (rectangles), square
numbers, cube numbers, triangle numbers (like a stair case).

Roadmap
-------

Some features I may add to this mod in future include:
 - Add tools to make things easier and faster, e.g.:
   - Extend the group of blocks in the chosen direction
   - Rearrange groups of blocks to explore factors and other properties
 - Selection of blocks in a group for counting or other operations, making them
   light up
 - Add some form of arithmetic operators to combine (add/multiply), split
   (subtract/divide), or otherwise manipulate (square/cube) groups of blocks
 - Sounds (reading out of numbers)


CHANGELOG.md
Changes in [0.4](https://github.com/amalon/minetest-numeracy/releases/tag/v0.4) (2021-11-18)
============================================================================================

 * New original square arrangements: 121, 144
 * Update square arrangements: 81
 * New rectangle arrangements: 32 (4x8)
 * New cube arrangement: 64
 * New triangle arrangements: 36, 45, 55
 * New irregular arrangements: 13, 17, 19 (monster), 31 (calendar),
   55 (rocket), 80 (roboctoblock, spidoctoblock, dinoctoblock)

Changes in [0.3](https://github.com/amalon/minetest-numeracy/releases/tag/v0.3) (2021-04-13)
============================================================================================

 * Fix for F-Droid Minetest (with Lua < 5.2)
 * Add square arrangements 25-81
 * Darken sides of 40, 50, 90, and all the hundreds blocks
 * Basic support for numbers 1000-10000

Changes in [0.2](https://github.com/amalon/minetest-numeracy/releases/tag/v0.2) (2021-04-11)
============================================================================================

 * Basic support for numbers 100-999
 * Stop overlap of 10 blocks
 * Several internal code cleanups

Changes in [0.1](https://github.com/amalon/minetest-numeracy/releases/tag/v0.1) (2021-04-04)
============================================================================================
Initial release


utils.lua
--[[
https://lua-users.org/wiki/SortedIteration
Ordered table iterator, allow to iterate on the natural order of the keys of a
table.
Modified for reverse sorting by James Hogan <james@albanarts.com>
]]

local function __gen_ordered_index( t, dir )
    local ordered_index = {}
    for key in pairs(t) do
        table.insert( ordered_index, key )
    end
    if dir > 0 then
        table.sort( ordered_index )
    else
        -- reverse sort
        table.sort( ordered_index, function(a,b)
            return a > b
        end)
    end
    return ordered_index
end

local function ordered_next(t, state, dir)
    -- Equivalent of the next function, but returns the keys in the alphabetic
    -- order. We use a temporary ordered key table that is stored in the
    -- table being iterated.

    local key = nil
    --print("ordered_next: state = "..tostring(state) )
    if state == nil then
        -- the first time, generate the index
        t.__ordered_index = __gen_ordered_index( t, dir or 1 )
        key = t.__ordered_index[1]
    else
        -- fetch the next value
        for i = 1,table.getn(t.__ordered_index) do
            if t.__ordered_index[i] == state then
                key = t.__ordered_index[i+1]
            end
        end
    end

    if key then
        return key, t[key]
    end

    -- no more value to return, cleanup
    t.__ordered_index = nil
    return
end

local function ordered_next_reverse(t, state)
	return ordered_next(t, state, -1)
end

function numeracy_ordered_pairs(t, dir)
    -- Equivalent of the pairs() function on tables. Allows to iterate
    -- in order
    if (dir or 1) >= 0 then
        return ordered_next, t, nil
    else
        return ordered_next_reverse, t, nil
    end
end


numbers.lua
-- shape of centred single digit numbers
local node_box_single_centre = {
	type = "fixed",
	fixed = {
		{-3/16, -6/16, -1/16,  3/16, 6/16, 1/16},
	},
}
-- shape of right justified single digit numbers
local node_box_single_right = {
	type = "fixed",
	fixed = {
		{1/16, -6/16, 0,  7/16, 6/16, 0},
	},
}
-- shape of double digit numbers
local node_box_double = {
	type = "fixed",
	fixed = {
		{-7/16, -6/16, 0,  7/16, 6/16, 0},
	},
}

for i = 1,9 do
	minetest.register_node("numeracy:number_centre_"..tostring(i), {
		description = "Number "..tostring(i).." (centre)",
		tiles = {
			"numeracy_blank.png",
			"numeracy_blank.png",
			"numeracy_blank.png",
			"numeracy_blank.png",
			"numeracy_blank.png^[combine:8x16:4,0=numeracy_number_"..tostring(i)..".png^[transformFX",
			"numeracy_blank.png^[combine:8x16:4,0=numeracy_number_"..tostring(i)..".png",
		},
		use_texture_alpha = 'clip',
		groups = { cracky = 2, not_in_creative_inventory = 1 },
		drop = "",

		pointable = false,
		walkable = false,

		paramtype = "light",
		paramtype2 = "facedir", -- the upper 3 bits represent direction numbers extend

		drawtype = "nodebox",
		node_box = node_box_single_centre,
	})
	minetest.register_node("numeracy:number_right_"..tostring(i), {
		description = "Number "..tostring(i).." (right)",
		tiles = {
			"numeracy_blank.png",
			"numeracy_blank.png",
			"numeracy_blank.png",
			"numeracy_blank.png",
			"numeracy_blank.png^[combine:8x16:8,0=numeracy_number_"..tostring(i)..".png^[transformFX",
			"numeracy_blank.png^[combine:8x16:8,0=numeracy_number_"..tostring(i)..".png",
		},
		use_texture_alpha = 'clip',
		groups = { cracky = 2, not_in_creative_inventory = 1 },
		drop = "",

		pointable = false,
		walkable = false,

		paramtype = "light",
		paramtype2 = "facedir", -- the upper 3 bits represent direction numbers extend

		drawtype = "nodebox",
		node_box = node_box_single_right,
	})
end
for i = 0,9 do
	for j = 0,9 do
		minetest.register_node("numeracy:number_"..tostring(i)..tostring(j), {
			description = "Number "..tostring(i)..tostring(j),
			tiles = {
				"numeracy_blank.png",
				"numeracy_blank.png",
				"numeracy_blank.png",
				"numeracy_blank.png",
				"numeracy_blank.png^[combine:8x16:0,0=numeracy_number_"..tostring(i)..".png:8,0=numeracy_number_"..tostring(j)..".png^[transformFX",
				"numeracy_blank.png^[combine:8x16:0,0=numeracy_number_"..tostring(i)..".png:8,0=numeracy_number_"..tostring(j)..".png",
			},
			use_texture_alpha = 'clip',
			groups = { cracky = 2, not_in_creative_inventory = 1 },
			drop = "",
			pointable = false,
			walkable = false,

			paramtype = "light",
			paramtype2 = "facedir", -- the upper 3 bits represent direction numbers extend

			drawtype = "nodebox",
			node_box = node_box_double,
		})
	end
end


blocks.lua
dofile(minetest.get_modpath("numeracy") .. "/update.lua")

-- shape of blocks, 1/16 shaved off along the edges
local node_box = {
	type = "fixed",
	fixed = {
		{-8/16, -7/16, -7/16,  8/16, 7/16, 7/16},
		{-7/16, -8/16, -7/16,  7/16, 8/16, 7/16},
		{-7/16, -7/16, -8/16,  7/16, 7/16, 8/16},
	},
}

local ten_min = 6/16
local ten_sur = 7/16
local ten_max = 8/16
local node_box_ten = {
	type = "connected",

	-- slightly small cube
	fixed = {
		{-ten_sur, -ten_sur, -ten_sur,   ten_sur,  ten_sur,  ten_sur},
	},

	-- extend sides towards neighbouring connections
	connect_top = {
		{-ten_sur,  ten_sur, -ten_sur,   ten_sur,  ten_max,  ten_sur},
	},
	connect_bottom = {
		{-ten_sur, -ten_max, -ten_sur,   ten_sur, -ten_sur,  ten_sur},
	},
	connect_left = {
		{-ten_max, -ten_sur, -ten_sur,  -ten_sur,  ten_sur,  ten_sur},
	},
	connect_right = {
		{ ten_sur, -ten_sur, -ten_sur,   ten_max,  ten_sur,  ten_sur},
	},
	connect_front = {
		{-ten_sur, -ten_sur, -ten_max,   ten_sur,  ten_sur, -ten_sur},
	},
	connect_back = {
		{-ten_sur, -ten_sur,  ten_sur,   ten_sur,  ten_sur,  ten_max},
	},

	-- extend unconnected faces tangentially to show a slightly ugly but
	-- functional border around the groups of blocks
	disconnected_top = {
		{-ten_max,  ten_min, -ten_max,   ten_max,  ten_sur,  ten_max},
	},
	disconnected_bottom = {
		{-ten_max, -ten_sur, -ten_max,   ten_max, -ten_min,  ten_max},
	},
	disconnected_left = {
		{-ten_sur, -ten_max, -ten_max,  -ten_min,  ten_max,  ten_max},
	},
	disconnected_right = {
		{ ten_min, -ten_max, -ten_max,   ten_sur,  ten_max,  ten_max},
	},
	disconnected_front = {
		{-ten_max, -ten_max, -ten_sur,   ten_max,  ten_max, -ten_min},
	},
	disconnected_back = {
		{-ten_max, -ten_max,  ten_min,   ten_max,  ten_max,  ten_sur},
	},
}

minetest.register_node("numeracy:block", {
	description = "Numeracy block",
	tiles = {
		"numeracy_block_white_side.png"
	},
	palette = "numeracy_block_palette.png",
	color = "#FF002B",
	groups = { cracky = 1 },
	drop = "numeracy:block",

	drawtype = "nodebox",
	paramtype = "light",
	paramtype2 = "color",
	node_box = node_box,

	on_place = numeracy_block_on_place,
	after_dig_node = numeracy_block_after_dig_node,
})

local ten_blocks = {
	[10] = {
		tile_side  = "numeracy_block_white_side.png^[multiply:#FF002B",
		tile_front = "numeracy_block_white_side.png"
	},
	[20] = {
		qty        = 2,
		tile_side  = "numeracy_block_white_side.png^[multiply:#FF9500",
		tile_front = "numeracy_block_white_side.png^[multiply:#F7D98D",
	},
	[30] = {
		qty        = 3,
		tile_side  = "numeracy_block_white_side.png^[multiply:#F4E41C",
		tile_front = "numeracy_block_white_side.png^[multiply:#FDFF9C",
	},
	[40] = {
		qty        = 4,
		tile_side  = "numeracy_block_white_side.png^[multiply:#6BDB63",
		tile_front = "numeracy_block_white_side.png^[multiply:#BEFFA6",
	},
	[50] = {
		qty        = 5,
		tile_side  = "numeracy_block_white_side.png^[multiply:#5AC7E4",
		tile_front = "numeracy_block_white_side.png^[multiply:#A9F5EA",
	},
	[60] = {
		qty        = 6,
		tile_side  = "numeracy_block_white_side.png^[multiply:#613999",
		tile_front = "numeracy_block_white_side.png^[multiply:#9870E1",
	},
	[71] = {
		tile_side  = "numeracy_block_white_side.png^[multiply:#FF002B",
		tile_front = "numeracy_block_white_side.png^[multiply:#FF8C9B",
	},
	[77] = {
		tile_side  = "numeracy_block_white_side.png^[multiply:#C777E5",
		tile_front = "numeracy_block_white_side.png^[multiply:#D0A0FF",
	},
	[80] = {
		qty        = 8,
		tile_side  = "numeracy_block_white_side.png^[multiply:#F73BAB",
		tile_front = "numeracy_block_white_side.png^[multiply:#FF96E0",
	},
	[90] = {
		qty        = 3,
		tile_side  = "numeracy_block_white_side.png^[multiply:#646567",
		tile_front = "numeracy_block_white_side.png^[multiply:#C8CBD0",
	},
	[91] = {
		qty        = 3,
		tile_side  = "numeracy_block_white_side.png^[multiply:#515253",
		tile_front = "numeracy_block_white_side.png^[multiply:#A3A4A6",
	},
	[92] = {
		qty        = 3,
		tile_side  = "numeracy_block_white_side.png^[multiply:#424746",
		tile_front = "numeracy_block_white_side.png^[multiply:#848F8B",
	},
	-- 10000s are rather similar to 10s
	[10000] = {
		qty        = 10,
		tile_side  = "numeracy_block_white_side.png^[multiply:#FF002B",
		tile_front = "numeracy_block_white_side.png"
	},
}

for i,info in pairs(ten_blocks) do
	local qty = info.qty or 1
	for j = 0,qty - 1 do
		minetest.register_node("numeracy:block_"..tostring(i).."_"..tostring(j), {
			description = "Numeracy block "..tostring(i).." ("..tostring(j)..")",
			tiles = {
				info.tile_side,
				info.tile_side,
				info.tile_side,
				info.tile_side,
				info.tile_front,
				info.tile_front,
			},
			groups = { cracky = 2, not_in_creative_inventory = 1 },
			drop = "numeracy:block",

			drawtype = "nodebox",
			node_box = node_box_ten,

			connects_to = { "numeracy:block_"..tostring(i).."_"..tostring(j) },

			paramtype = "light",
			paramtype2 = "facedir",

			on_place = numeracy_block_on_place,
			after_dig_node = numeracy_block_after_dig_node,
		})
	end
end

local hundred_blocks = {
	[100] = {
		palette    = "numeracy_block_100_palette.png",
		tile_side  = "numeracy_block_white_side.png^[multiply:#FF002B",
		tile_front = "numeracy_block_white_side.png",
	},
	[200] = {
		qty        = 2,
		palette    = "numeracy_block_100_palette.png",
		tile_side  = "numeracy_block_white_side.png^[multiply:#FF9500",
		tile_front = "numeracy_block_white_side.png",
	},
	[300] = {
		qty        = 3,
		palette    = "numeracy_block_100_palette.png",
		tile_side  = "numeracy_block_white_side.png^[multiply:#F4E41C",
		tile_front = "numeracy_block_white_side.png",
	},
	[400] = {
		qty        = 4,
		palette    = "numeracy_block_100_palette.png",
		tile_side  = "numeracy_block_white_side.png^[multiply:#6BDB63",
		tile_front = "numeracy_block_white_side.png",
	},
	[500] = {
		qty        = 5,
		palette    = "numeracy_block_500_palette.png",
		tile_side  = "numeracy_block_white_side.png^[multiply:#5AC7E4",
		tile_front = "numeracy_block_white_side.png",
	},
	[600] = {
		qty        = 6,
		palette    = "numeracy_block_500_palette.png",
		tile_side  = "numeracy_block_white_side.png^[multiply:#613999",
		tile_front = "numeracy_block_white_side.png",
	},
	[700] = {
		palette    = "numeracy_block_500_palette.png",
		tile_side  = "numeracy_block_white_side.png^[multiply:#C777E5",
		tile_front = "numeracy_block_white_side.png",
	},
	[800] = {
		qty        = 8,
		palette    = "numeracy_block_500_palette.png",
		tile_side  = "numeracy_block_white_side.png^[multiply:#F73BAB",
		tile_front = "numeracy_block_white_side.png",
	},
	[900] = {
		qty        = 3,
		palette    = "numeracy_block_900_palette.png",
		tile_side  = "numeracy_block_white_side.png^[multiply:#7f7f7f",
		tile_front = "numeracy_block_white_side.png",
	},
	[901] = {
		qty        = 3,
		palette    = "numeracy_block_900_palette.png",
		tile_side  = "numeracy_block_white_side.png^[multiply:#7f7f7f",
		tile_front = "numeracy_block_white_side.png",
	},
	[902] = {
		qty        = 3,
		palette    = "numeracy_block_900_palette.png",
		tile_side  = "numeracy_block_white_side.png^[multiply:#7f7f7f",
		tile_front = "numeracy_block_white_side.png",
	},
}

for i,info in pairs(hundred_blocks) do
	local qty = info.qty or 1
	for j = 0,qty - 1 do
		minetest.register_node("numeracy:block_"..tostring(i).."_"..tostring(j), {
			description = "Numeracy block "..tostring(i).." ("..tostring(j)..")",
			tiles = {
				info.tile_side,
				info.tile_side,
				info.tile_side,
				info.tile_side,
				info.tile_front,
				info.tile_front,
			},
			palette = info.palette,
			groups = { cracky = 2, not_in_creative_inventory = 1 },
			drop = "numeracy:block",

			drawtype = "nodebox",
			node_box = node_box_ten,

			connects_to = { "numeracy:block_"..tostring(i).."_"..tostring(j) },

			paramtype = "light",
			paramtype2 = "colorfacedir",

			on_place = numeracy_block_on_place,
			after_dig_node = numeracy_block_after_dig_node,
		})
	end
end

local thousand_blocks = {
	[1000] = {},
	[2000] = {
		qty        = 2,
	},
	[3000] = {
		qty        = 3,
	},
	[4000] = {
		qty        = 4,
	},
	[5000] = {
		qty        = 5,
	},
	[6000] = {
		qty        = 6,
	},
	[7000] = {},
	[8000] = {
		qty        = 8,
	},
	[9000] = {
		qty        = 3,
	},
	[9001] = {
		qty        = 3,
	},
	[9002] = {
		qty        = 3,
	},
}

for i,info in pairs(thousand_blocks) do
	local qty = info.qty or 1
	for j = 0,qty - 1 do
		minetest.register_node("numeracy:block_"..tostring(i).."_"..tostring(j), {
			description = "Numeracy block "..tostring(i).." ("..tostring(j)..")",
			tiles = { "numeracy_block_white_side.png" },
			palette = "numeracy_block_palette.png",
			groups = { cracky = 2, not_in_creative_inventory = 1 },
			drop = "numeracy:block",

			drawtype = "nodebox",
			node_box = node_box_ten,

			connects_to = { "numeracy:block_"..tostring(i).."_"..tostring(j) },

			paramtype = "light",
			paramtype2 = "color",

			on_place = numeracy_block_on_place,
			after_dig_node = numeracy_block_after_dig_node,
		})
	end
end
